### PR TITLE
Support partial mempool synchronization

### PR DESCRIPTION
**Pull request by shesek**
Prior to this change, the local electrs mempool was only updated with new transactions once we were able to get a complete snapshot of every transaction in bitcoind's mempool, which could require multiple rounds of attempts if any transactions are replaced (or otherwise evicted) between querying for the mempool txids and querying for the transactions themselves.

PR #89 made each round more efficient, but obtaining a complete snapshot can still potentially take many rounds the more RBF is used, with each round taking longer the larger the mempool gets (just listing the mempool txids to find the delta becomes costly).

With this change, the local mempool is instead updated with whatever transactions we are able to get, without waiting for a fully consistent snapshot - which should improve performance and reduce latency.

Making partial updates requires some extra care on the electrs side to ensure mempool consistency (i.e., no double-spends or missing input txos), since we can no longer fully rely on bitcoind validating consistency for us. Specifically:

- If any mempool transactions became unavailable (replaced) by the time we queried for them, we explicitly check for transactions that depend on the missing transactions (or their descendants) and remove them too.

- Mempool evictions are processed first before any additions, even if the replacing transactions aren't available. This ensures double-spend conflicts are not possible. (technically this was already the case, but now it became more critical)

### CODE REVIEW

**Comment by shesek** [comment]
This could definitely use some tests, but testing the edge-cases here is pretty tricky (in particular, the case where we manage to get the child tx before the parent is RBF'd, but then miss the parent). Still thinking on ways to approach it.

File: src/new_index/mempool.rs [hunk 1]
```diff
@@ -516,99 +516,113 @@ impl Mempool {
...
+                replaced_txids.len(),
+                replaced_txids.iter().take(10).collect::<Vec<_>>()
+            );
+        }
+
+        // If we were unable to get a complete consistent snapshot of the bitcoind mempool,
+        // detect and remove any transactions that spend from the missing (replaced) transactions
+        // or any of their descendants. This is necessary because it could be possible to fetch the
+        // child tx successfully before the parent is replaced, but miss the replaced parent tx.
+        while !replaced_txids.is_empty() {
```
Comment (Randy808) [inline src/new_index/mempool.rs:598]: Is this descendant cleanup needed? Won't these transactions get removed in 
the next update() call (within 5 seconds) when their missing parents aren't 
in bitcoind's mempool txid list?

Comment (Randy808) [reply src/new_index/mempool.rs:598]: I think you addressed this in the PR description:
"Making partial updates requires some extra care on the electrs side to ensure mempool consistency (i.e., no double-spends or missing input txos)"

The objective is to have no references to parent transactions we aren't able to retrieve

**Review by Randy808 — APPROVED** [review]

**Comment by philippem** [comment]
tack

**Review by philippem — APPROVED** [review]

File: src/new_index/mempool.rs [hunk 2]
```diff
@@ -516,99 +516,113 @@ impl Mempool {
...
-            // Abort if the chain tip moved while fetching transactions
-            if daemon.getbestblockhash()? != *tip {
-                warn!("chain tip moved while updating mempool");
-                return Ok(false);
-            }
+        // Find transactions available in bitcoind's mempool but not indexed locally
+        let new_txids = bitcoind_txids
+            .iter()
+            .filter(|&txid| !indexed_txids.contains(txid))
+            .collect::<Vec<_>>();
```
Comment (RCasatta) [inline src/new_index/mempool.rs:544]: nit: shouldn't this be another difference() ?

**Comment by RCasatta** [comment]
> just listing the mempool txids to find the delta becomes costly

just mentioning it is usually pretty fast, I am doing similar logic in another repo

image

image

**Review by RCasatta — DISMISSED** [review]
utACK 3ff21568f21a68f893e64dcba8436ad909d3a71f

**Review by philippem — DISMISSED** [review]

File: src/new_index/mempool.rs (hunk 2 above)
Comment (philippem) [reply src/new_index/mempool.rs:544]: @shesek ?

Comment (shesek) [reply src/new_index/mempool.rs:544]: Yes you're absolutely right, thanks. It ended up like this because it was rewritten from the [previous code](https://github.com/Blockstream/electrs/pull/168/files#diff-50b7ebc43d9c768232bcb93c081a3cbb4471b81f41e10a88e561f2a189c6c954L545-L548) that checked against two `HashSet`s.

Changed to `difference()` in c3ce3a96b0434cab16457b4bd62ac20277ddd423.

**Review by philippem — APPROVED** [review]
